src/bash/plugin.c

```diff
@@ -0,0 +1,394 @@
+            /* ignore comments or white space. */
+            continue; 
+        }
+        
+        /* read to first whitespace. */
+        strtok(buffer, " \t\n\r\f"); 
+
+        if(!strncmp(buffer, "plugin=", 7)) {
+            /* read plugin path. */
+            char* plugin_path = strtok(buffer+7, " \t\n\r\f"); 
+#ifdef DEBUG
+            itrace("Plugin: load plugin: %s\n", plugin_path);
+#endif
+            try_load_plugin_by_path(plugin_path);
+        }
+#ifdef DEBUG
+        else {
+            /* output debug message. */
+            itrace("Plugin: unrecognized parameter: %s\n", buffer);
+        }
+#endif
+    }
+
+    fclose(config_file);
+}
+
+/* Free loaded plugins */
+void
+free_loaded_plugins()
+{
+    if (global_plugin_list == NULL) {
+        return;
+    }
+	
+#ifdef DEBUG
+    itrace("Plugin: start free plugin from global list %p\n", global_plugin_list);
+#endif
+
+    /* Walk to last plugin */
+    PLUGIN_NODE *next_plugin_node = global_plugin_list;
+    while (next_plugin_node != NULL) {
+
+        /* Unload plugin */
+        next_plugin_node->plugin_uninit();
+        dlclose(next_plugin_node->plugin_handle);
+
+        /* Continue with next pligin */
+		PLUGIN_NODE* current_plugin_node_memory = next_plugin_node;
+        next_plugin_node = next_plugin_node->next;
+
+#ifdef DEBUG
+		itrace("Plugin: next plugin address %p\n", next_plugin_node);
+#endif
+
+		/* Free plugin node memory, this may also reset all allocated memory depends on c lib implementation */
+		free(current_plugin_node_memory);
+    }
+
+    /* Reset plugin list */
+	global_plugin_list = NULL;
+}
+
+/* Invoke loaded plugins */
+int
+invoke_loaded_plugins (user, shell_level, cmd, argv)
+     char *user;
+     int shell_level;
+     char *cmd;
+     char **argv;
+{
+    if (global_plugin_list == NULL) {
+        return 0;
+    }
+	
+#ifdef DEBUG
+    itrace("Plugin: start invoke plugin from global list %p\n", global_plugin_list);
+#endif
+
+    /* Walk to last plugin */
+    PLUGIN_NODE *next_plugin_node = global_plugin_list;
+    while (next_plugin_node != NULL) {
+
+        /* Call plugin method */
+        int plugin_error_code = next_plugin_node->on_shell_execve(user, shell_level, cmd, argv);
+        if (plugin_error_code != 0) {
+#ifdef DEBUG
+            itrace("Plugin: on_execve return error: %d\n", plugin_error_code);
+#endif
+            /* Exit when plugin failed */
+            return plugin_error_code;
+        }
+        
+        /* Continue with next pligin */
+        next_plugin_node = next_plugin_node->next;
+		
+#ifdef DEBUG
+		itrace("Plugin: next plugin address %p\n", next_plugin_node);
+#endif
+    }
+    
+    return 0;
+}
+
+/* Load all plugins。 */
+void
+load_plugins ()
+{
+    load_plugin_by_config(plugin_config_file);
+}
+
+/* Free all plugins */
+void
+free_plugins ()
+{
+    free_loaded_plugins();
+}
+
+/* Invoke plugins before shell execve */
+int
+invoke_plugin_on_shell_execve (user, cmd, argv)
+     char *user;
```

Comment:
>      [](http://example.com/codeflow?start=0&length=5)

Is 5 space indentation a style for bash code?

Reply:
Yes, other code seems all following this style.